sys/arm64/arm64/pmap.c

```diff
@@ -1231,7 +1231,7 @@ pmap_bootstrap_dmap(vm_paddr_t min_pa)
 
 #ifdef __CHERI_PURE_CAPABILITY__
 	dmap_base_cap = cheri_setaddress(kernel_root_cap, DMAP_MIN_ADDRESS);
-	dmap_base_cap = cheri_setbounds(dmap_base_cap,
+	dmap_base_cap = cheri_setboundsexact(dmap_base_cap,
```

Comment:
I'm actually quite skeptical that this will be representable in all cases.  dmap_phys_base and dmap_phys_max are the smallest and largest RAM address and you have no control over those to assume that the distance between those will be on a representable boundary.  Certainly if you imagine a typical X86 system which has an odd 512MB hole at 3.5GB PA to support PCI mmio (as well as the 1MB ISA hole at 15MB PA), you will not have the highest PA of an N GB system be N GB, it's instead N GB + 512M + 1M.  arm64 is potentially worse as it can have sparse PA with large holes between RAM-backed PA ranges rather than the mostly dense layout on X86.

Reply:
Right. It is a bit sad that this can be unrepresentable in some cases but everything goes through PHYS_IN_DMAP so perhaps it's not terrible either. The pedantic course of action would be to get guard pages into the padding space but I'll revert this for now.

Reply:
So the virtual address space reserved for the DMAP is only ever used for the DMAP and nothing else, there's no danger of anything being mapped underneath the "tail" at the end of the capability, so long as the statically allocated VA reservation for the DMAP is itself representable.  For example from RISC-V's <machine/vmparam.h>:

```
 * SV39 memory map:
 * 0x0000000000000000 - 0x0000003fffffffff    256GB user map
 * 0x0000004000000000 - 0xffffffbfffffffff    unmappable
 * 0xffffffc000000000 - 0xffffffc7ffffffff    32GB kernel map
 * 0xffffffc800000000 - 0xffffffcfffffffff    32GB unused
 * 0xffffffd000000000 - 0xffffffefffffffff    128GB direct map
 * 0xfffffff000000000 - 0xffffffffffffffff    64GB unused
 *
 * SV48 memory map:
 * 0x0000000000000000 - 0x00007fffffffffff    128TB user map
 * 0x0000800000000000 - 0xffff7fffffffffff    unmappable
 * 0xffff800000000000 - 0xffffffc7ffffffff    127.75TB hole
 * 0xffffffc000000000 - 0xffffffc7ffffffff    32GB kernel map
 * 0xffffffc800000000 - 0xffffffcfffffffff    32GB unused
 * 0xffffffd000000000 - 0xffffffefffffffff    128GB direct map
 * 0xfffffff000000000 - 0xffffffffffffffff    64GB unused
```

Aarch64 is similar:

```
 * We use the full 48 bits for each region, however the kernel may only use
 * a limited range within this space.
 *
 * Upper region:    0xffffffffffffffff  Top of virtual memory
 *
 *                  0xfffffeffffffffff  End of DMAP
 *                  0xfffffa0000000000  Start of DMAP
 *
 *                  0xffff007fffffffff  End of KVA
 *                  0xffff000000000000  Kernel base address & start of KVA
 *
 * Hole:            0xfffeffffffffffff
 *                  0x0001000000000000
 *
 * Lower region:    0x0000ffffffffffff End of user address space
 *                  0x0000000000000000 Start of user address space
```

Reply:
Thanks for the clarification, I have a new set of updates that reverts this.